keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -843,24 +992,57 @@ examples:
 
 ### Termination of containers
 
-In alpha sidecar containers will be terminated as regular containers. No special
+In Alpha sidecar containers will be terminated as regular containers. No special
 or additional signals will be supported.
 
-We will collect feedback from alpha implementation and plan to improve
-termination in Beta. When Pods with sidecars are terminated:
+In Beta we have thought about introducing additional termination grace period fields
+to manage termination duration
+([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
+and leverage these fields to add reverse order termination of sidecar containers
+after the primary containers terminate.
 
+However, we decided on an alternative that doesn't require additional fields or hooks while keeping
+the desired behaviors when Pods with sidecars are terminated. While original approach works better
+with truly graceful termination where consistency is more important than time taken, proposed approach
+works for that scenario as well as a more and more popular scenario of limited time to terminate when
+graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
+as much as possible (think of a Spot Instances with 30 seconds notification).
+
+Here is the proposed approach:
+1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
+   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
+   the Pod's termination grace period countdown.
+2. Sidecar containers enter the `Terminated` state and are no longer restarted if they fail.
```

Comment:
Is this a feature ("we don't want them to restart") or a lack of feature ("we would prefer they restart but it's hard") ?

We still (TTBOMK) don't have a good state diagram for pod lifecycle.  This seems like it could be a problem.

E.g. if I have a 60m TGPS and my logging sidecar dies, no more logs?

Reply:
yes, this is what will happen. The problem is that it is close to impossible to distinguish desired termination vs. crash.

Reply:
> The problem is that it is close to impossible to distinguish desired termination vs. crash.

I don't understand what you mean.  If `deletionTimestamp` is set it's a termination, if not it's a crash?

Reply:
I am talking about a sidecar. Once deletionTimestamp is set, we call `preStop`. Sidecar has an option to exit on a spot. And restarting it after this may not be what user will desire. However we cannot distinguish if container exit by itself or crashed.

Reply:
Make it `OnFailure` ?  Frankly, "restartPolicy=Always" was probably a mistake.  We should have only had "Never" and "OnFailure".

Reply:
xref https://github.com/kubernetes/enhancements/pull/4183#discussion_r1342069396 - I think it's the same point.

Reply:
Bringing my comment up here. Same issue on [Agones](https://agones.dev/) we run a sidecar SDK server that is connected to over localhost, ideally we would want the sidecar to restart while within a TGPS timeframe.

> The problem is that it is close to impossible to distinguish desired termination vs. crash.

Probably a naive question (with no idea how kubelet etc work), but do exit codes help here?

Reply:
@markmandel I think `OnFailure` is the exit code based solution - and I like the idea that sidecars restart by default `OnFailure`.

Reply:
restarting past the deletion timestamp on failure sounds like a good addition. However it complicates things a bit - do we call `preStop` right after start on every sidecar restart?

Reply:
...and TIL what `OnFailure` is. Honestly, wasn't even aware that was a thing! 😄 

In which case +1 on a restart policy `OnFailure` as the option during TGPS timeframe. That seems ideal to me, and also gives enough control to the end user.

Ignore if already been covered, but also floating the idea of giving the end user the option of what restart policy they want during TGPS for a sidecar? I could also understand if that's way too much complexity for container/pod lifecycle.

Reply:
> Ignore if already been covered, but also floating the idea of giving the end user the option of what restart policy they want during TGPS for a sidecar? I could also understand if that's way too much complexity for container/pod lifecycle.

we don't support this for regular containers. I'd say if we need it, we will need to do a bigger change that is not necessarily bound to sidecars.

We want to preserve the behavior of TGPS - it should be respected no matter whether sidecars are there or not. And we also think that if we will start taking parts of TGPS for sidecars exclusively, than back compat will break causing more problems.

Reply:
That's fair. Just figured I'd float the idea to see if it stuck in any way.

Reply:
> Once deletionTimestamp is set, we call preStop. Sidecar has an option to exit on a spot ... we cannot distinguish if container exit by itself or crashed.

I disagree - `preStop` is NOT a signal that you should exit, it's a signal that your time to exit is imminent.  Hence "pre" in the name.  If we call `preStop` and then you exit, we _should_ restart you EVEN IF YOU EXIT(0).  We did not call SIGTERM yet, so you are still subject to restarts.  We don't have a "terminating" state, per se, but deletionTimestamp is insufficient.

Reply:
I agree with @thockin. I'm not sure about the specific use case for terminating a process on its own within the PreStop Hook. Is this related to "Abbreviated termination" scenario? If so, should the sidecar container exit the PreStop Hook as early as possible and just let SIGTERM to terminate the process?